src/sources/host_metrics.rs

```diff
@@ -0,0 +1,1189 @@
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "host_metrics")]
+impl SourceConfig for HostMetricsConfig {
+    async fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        shutdown: ShutdownSignal,
+        out: Pipeline,
+    ) -> crate::Result<super::Source> {
+        Ok(Box::new(self.clone().run(out, shutdown).boxed().compat()))
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn source_type(&self) -> &'static str {
+        "host_metrics"
+    }
+}
+
+macro_rules! tags {
+    ( $( $key:expr => $value:expr ),* ) => {{
+        #[allow(unused_mut)]
+        let mut result = std::collections::BTreeMap::default();
+        $( result.insert($key.to_string(), $value.to_string()); )*
+            result
+    }}
+}
+
+impl HostMetricsConfig {
+    async fn run(self, mut out: Pipeline, shutdown: ShutdownSignal) -> Result<(), ()> {
+        let interval = Duration::from_secs(self.scrape_interval_secs);
+        let mut interval = time::interval(interval).map(|_| ());
+        let mut shutdown = shutdown.compat();
+
+        loop {
+            select! {
+                Some(()) = interval.next() => (),
+                _ = &mut shutdown => break,
+                else => break,
+            };
+
+            let metrics = self.capture_metrics().await;
+
+            let (sink, _) = out
+                .send_all(futures01::stream::iter_ok(metrics))
+                .compat()
+                .await
+                .map_err(|error| error!(message = "Error sending host metrics", %error))?;
+            out = sink;
+        }
+
+        Ok(())
+    }
+
+    fn has_collector(&self, collector: Collector) -> bool {
+        match &self.collectors {
+            None => true,
+            Some(collectors) => collectors.iter().any(|&c| c == collector),
+        }
+    }
+
+    async fn capture_metrics(&self) -> impl Iterator<Item = Event> {
+        let hostname = crate::get_hostname();
+        let mut metrics = Vec::new();
+        if self.has_collector(Collector::Cpu) {
+            metrics.extend(add_collector("cpu", self.cpu_metrics().await));
+        }
+        if self.has_collector(Collector::Disk) {
+            metrics.extend(add_collector("disk", self.disk_metrics().await));
+        }
+        if self.has_collector(Collector::Filesystem) {
+            metrics.extend(add_collector("filesystem", self.filesystem_metrics().await));
+        }
+        if self.has_collector(Collector::Load) {
+            metrics.extend(add_collector("load", self.loadavg_metrics().await));
+        }
+        if self.has_collector(Collector::Memory) {
+            metrics.extend(add_collector("memory", self.memory_metrics().await));
+            metrics.extend(add_collector("memory", self.swap_metrics().await));
+        }
+        if self.has_collector(Collector::Network) {
+            metrics.extend(add_collector("network", self.network_metrics().await));
+        }
```

Comment:
It's probably not worth changing now, but as we add more of these we could think about moving these collection methods onto the collectors themselves, so you could do something like

```rust
for collector in self.collectors {
    metrics.extend(collector.collect().await);
}
```

That'd also let us be a bit more modular as we implement more types of metrics.

Reply:
How would that work? `collectors` is `Vec<enum Collector>`, can you define methods on individual variants of an enum?

Reply:
It involve switching from an enum to a trait with `typetag` so that it can be deserialized. Similar to what we do with sources/sinks/transforms. Most of the enum solutions I can think of involve some degree of boilerplate.